torch/csrc/distributed/c10d/reducer.cpp

```diff
@@ -730,6 +753,59 @@ void Reducer::finalize_backward() {
     local_used_work_->wait();
   }
   local_used_maps_reduced_ = false;
+
+  // Rebuilt bucket if this is the first time to rebuilt
+  if (!rebuilt_tensors_.empty()) {
+    rebuildBuckets();
+  }
+}
+
+void Reducer::rebuildBuckets() {
+  TORCH_INTERNAL_ASSERT(
+      rebuilt_tensors_.size() == rebuilt_param_indices_.size(),
+      "rebuild tensor size is not same as rebuild param indices size.");
+  std::vector<std::vector<size_t>> rebuilt_bucket_indices;
+  std::vector<size_t> bucket_size_limits;
+  bucket_size_limits.push_back(DEFAULT_FIRST_BUCKET_BYTES);
+  bucket_size_limits.push_back(bucket_bytes_cap_);
+  rebuilt_bucket_indices = compute_bucket_assignment_by_size(
+      rebuilt_tensors_,
+      bucket_size_limits,
+      expect_sparse_gradients_[0],
+      rebuilt_param_indices_);
+
+  // For rebuilt bucket indices, it needs to be synced across all ranks.
+  // broadcast the newly rebuilt bucket indices from rank 0 in default.
+  // After sync rebuilt bucket indices, initialize buckets for reducer.
+  int64_t broadcast_bucket_size = DEFAULT_BROADCAST_BUCKET_BYTES;
+  auto num_buckets = rebuilt_bucket_indices.size();
+  auto num_elem_per_bucket = rebuilt_bucket_indices[0].size();
+  int64_t total_size = num_buckets * num_elem_per_bucket;
```

Comment:
is it guaranteed that all buckets have the same number of elements?

Reply:
Good catch, let me double check